File: Leetcode/C++/49. Group Anagrams/solution.cpp
// https://leetcode.com/problems/group-anagrams/

#include <vector>
#include <string>
#include <map>
#include <algorithm>

using namespace std;

class Solution
{
public:
    vector<vector<string>> groupAnagrams(vector<string> &strs)
    {
        vector<vector<string>> result;
        map<string, vector<string>> group;
        for (string s : strs)
        {
            string key = s;
            sort(key.begin(), key.end());
            group[key].push_back(s);
        }

        for (auto value : group)
        {
            result.push_back(value.second);
        }
        return result;
    }
};

File: Leetcode/Java/49. Group Anagrams/solution.java
// https://leetcode.com/problems/group-anagrams/description/

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> wordHash = new HashMap<>();

        for (String word: strs){
            char[] chars = word.toCharArray();
            Arrays.sort(chars);
            String key = String.valueOf(chars);

            if (!wordHash.containsKey(key)){
                wordHash.put(key, new ArrayList<>());
            }
            wordHash.get(key).add(String.valueOf(word));
        }

        return new ArrayList<>(wordHash.values()); 
    }
}